File: src/hynous/intelligence/tools/clusters.py
"""
Cluster Management Tool — manage_clusters

Lets the agent organize memories into named clusters for scoped search
and knowledge organization.

Actions:
  list    — List all clusters with node counts
  create  — Create a new cluster
  update  — Update cluster properties
  delete  — Delete a cluster
  add     — Add node(s) to a cluster
  remove  — Remove a node from a cluster
  members — List members of a cluster
  health  — Get health stats for a cluster

Standard tool module pattern:
  1. TOOL_DEF dict
  2. handler function
  3. register() wires into registry
"""

import json
import logging
from typing import Optional

logger = logging.getLogger(__name__)


TOOL_DEF = {
    "name": "manage_clusters",
    "description": (
        "Organize your memories into clusters for scoped search and "
        "knowledge management.\n\n"
        "Clusters group related memories — by asset (BTC, ETH), strategy "
        "(momentum, mean-reversion), market regime (bull, bear), or any "
        "category you choose.\n\n"
        "Actions:\n"
        "  list — List all clusters with node counts.\n"
        "  create — Create a new cluster. Optionally set auto_subtypes "
        "to auto-assign future memories.\n"
        "  update — Update cluster name, description, icon, pinned, "
        "or auto_subtypes.\n"
        "  delete — Delete a cluster (memories are NOT deleted, only "
        "the grouping).\n"
        "  add — Add memory node(s) to a cluster.\n"
        "  remove — Remove a memory from a cluster.\n"
        "  members — List memories in a cluster.\n"
        "  health — Show cluster health (active/weak/dormant counts).\n\n"
        "Auto-assignment: When you create a cluster with auto_subtypes "
        "(e.g. [\"thesis\", \"lesson\"]), all future memories of those "
        "types are automatically added to the cluster. A memory can "
        "belong to multiple clusters.\n\n"
        "Examples:\n"
        '  {"action": "list"}\n'
        '  {"action": "create", "name": "BTC Analysis", '
        '"description": "All BTC-related memories", '
        '"auto_subtypes": ["thesis", "lesson", "signal"]}\n'
        '  {"action": "add", "cluster_id": "cl_abc123", '
        '"node_id": "n_def456"}\n'
        '  {"action": "add", "cluster_id": "cl_abc123", '
        '"node_ids": ["n_def456", "n_ghi789"]}\n'
        '  {"action": "members", "cluster_id": "cl_abc123"}\n'
        '  {"action": "health", "cluster_id": "cl_abc123"}'
    ),
    "parameters": {
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": [
                    "list", "create", "update", "delete",
                    "add", "remove", "members", "health",
                ],
                "description": "What to do.",
            },
            "cluster_id": {
                "type": "string",
                "description": (
                    "Cluster ID (for update, delete, add, remove, "
                    "members, health actions)."
                ),
            },
            "name": {
                "type": "string",
                "description": "Cluster name (for create/update).",
            },
            "description": {
                "type": "string",
                "description": "Cluster description (for create/update).",
            },
            "icon": {
                "type": "string",
                "description": "Emoji or icon (for create/update).",
            },
            "pinned": {
                "type": "boolean",
                "description": "Pin cluster to top of list (for create/update).",
            },
            "auto_subtypes": {
                "type": "array",
                "items": {"type": "string"},
                "description": (
                    "Memory types to auto-assign to this cluster. "
                    "Use short names: thesis, lesson, signal, episode, "
                    "trade, curiosity, watchpoint, trade_entry, "
                    "trade_close, trade_modify."
                ),
            },
            "node_id": {
                "type": "string",
                "description": "Memory ID to add/remove (for add/remove actions).",
            },
            "node_ids": {
                "type": "array",
                "items": {"type": "string"},
                "description": "Multiple memory IDs to add (for add action).",
            },
            "limit": {
                "type": "integer",
                "description": "Max results for members action. Default 50.",
            },
        },
        "required": ["action"],
    },
}

# Short name → full subtype mapping (same subtypes as memory.py _TYPE_MAP)
_SUBTYPE_MAP = {
    "watchpoint": "custom:watchpoint",
    "curiosity": "custom:curiosity",
    "lesson": "custom:lesson",
    "thesis": "custom:thesis",
    "episode": "custom:market_event",
    "trade": "custom:trade",
    "signal": "custom:signal",
    "trade_entry": "custom:trade_entry",
    "trade_modify": "custom:trade_modify",
    "trade_close": "custom:trade_close",
    "turn_summary": "custom:turn_summary",
    "session_summary": "custom:session_summary",
}


def _normalize_subtypes(raw: list[str]) -> list[str]:
    """Convert short names to full custom: subtypes."""
    normalized = []
    for st in raw:
        if st in _SUBTYPE_MAP:
            normalized.append(_SUBTYPE_MAP[st])
        elif st.startswith("custom:"):
            normalized.append(st)
        else:
            # Unknown short name — prepend custom: as fallback
            normalized.append(f"custom:{st}")
    return normalized


def _format_subtypes_display(subtypes: list) -> str:
    """Format subtypes for display — strip custom: prefix."""
    return ", ".join(
        s.replace("custom:", "") if isinstance(s, str) else str(s)
        for s in subtypes
    )


def handle_manage_clusters(
    action: str,
    cluster_id: Optional[str] = None,
    name: Optional[str] = None,
    description: Optional[str] = None,
    icon: Optional[str] = None,
    pinned: Optional[bool] = None,
    auto_subtypes: Optional[list] = None,
    node_id: Optional[str] = None,
    node_ids: Optional[list] = None,
    limit: int = 50,
) -> str:
    """Handle all cluster management actions."""
    from ...nous.client import get_client

    try:
        client = get_client()

        # ---- LIST ----
        if action == "list":
            clusters = client.list_clusters()

            if not clusters:
                return "No clusters. Use create action to make one."

            lines = [f"{len(clusters)} cluster(s):", ""]

            for cl in clusters:
                cid = cl.get("id", "?")
                cname = cl.get("name", "Untitled")
                count = cl.get("node_count", 0)
                is_pinned = cl.get("pinned", 0)
                desc = cl.get("description", "")
                auto_subs = cl.get("auto_subtypes")

                pin_tag = " [pinned]" if is_pinned else ""
                lines.append(f"  {cname} ({cid}) — {count} nodes{pin_tag}")

                if desc:
                    lines.append(f"    {desc}")

                if auto_subs and isinstance(auto_subs, list) and len(auto_subs) > 0:
                    lines.append(
                        f"    Auto-assigns: {_format_subtypes_display(auto_subs)}"
                    )

                lines.append("")

            return "\n".join(lines)

        # ---- CREATE ----
        elif action == "create":
            if not name:
                return "Error: name is required for create action."

            # Normalize auto_subtypes to full form
            normalized_subs = None
            if auto_subtypes:
                normalized_subs = _normalize_subtypes(auto_subtypes)

            result = client.create_cluster(
                name=name,
                description=description,
                icon=icon,
                pinned=pinned or False,
                auto_subtypes=normalized_subs,
            )

            cid = result.get("id", "?")
            logger.info("Created cluster: \"%s\" (%s)", name, cid)

            lines = [f"Created cluster: \"{name}\" ({cid})"]
            if normalized_subs:
                lines.append(
                    f"Auto-assigns: {_format_subtypes_display(normalized_subs)}"
                )
            return "\n".join(lines)

        # ---- UPDATE ----
        elif action == "update":
            if not cluster_id:
                return "Error: cluster_id is required for update action."

            kwargs: dict = {}
            if name is not None:
                kwargs["name"] = name
            if description is not None:
                kwargs["description"] = description
            if icon is not None:
                kwargs["icon"] = icon
            if pinned is not None:
                kwargs["pinned"] = pinned
            if auto_subtypes is not None:
                kwargs["auto_subtypes"] = _normalize_subtypes(auto_subtypes)

            if not kwargs:
                return "Error: provide at least one field to update."

            result = client.update_cluster(cluster_id, **kwargs)
            updated_name = result.get("name", cluster_id)

            changed = list(kwargs.keys())
            logger.info(
                "Updated cluster: \"%s\" (%s) — %s",
                updated_name, cluster_id, ", ".join(changed),
            )
            return (
                f"Updated: \"{updated_name}\" ({cluster_id}) "
                f"— {', '.join(changed)}"
            )

        # ---- DELETE ----
        elif action == "delete":
            if not cluster_id:
                return "Error: cluster_id is required for delete action."

            client.delete_cluster(cluster_id)
            logger.info("Deleted cluster: %s", cluster_id)
            return (
                f"Deleted cluster {cluster_id}. "
                "Memories are NOT deleted — only the grouping is removed."
            )

        # ---- ADD ----
        elif action == "add":
            if not cluster_id:
                return "Error: cluster_id is required for add action."
            if not node_id and not node_ids:
                return "Error: node_id or node_ids is required for add action."

            result = client.add_to_cluster(
                cluster_id, node_id=node_id, node_ids=node_ids,
            )
            added = result.get("added", 0)
            total = len(node_ids) if node_ids else 1

            logger.info(
                "Added %d node(s) to cluster %s", added, cluster_id,
            )
            return f"Added {added}/{total} node(s) to cluster {cluster_id}."

        # ---- REMOVE ----
        elif action == "remove":
            if not cluster_id:
                return "Error: cluster_id is required for remove action."
            if not node_id:
                return "Error: node_id is required for remove action."

            client.remove_from_cluster(cluster_id, node_id)
            logger.info(
                "Removed %s from cluster %s", node_id, cluster_id,
            )
            return f"Removed {node_id} from cluster {cluster_id}."

        # ---- MEMBERS ----
        elif action == "members":
            if not cluster_id:
                return "Error: cluster_id is required for members action."

            members = client.get_cluster_members(cluster_id, limit=limit)

            if not members:
                return f"Cluster {cluster_id} has no members."

            lines = [f"Cluster {cluster_id} — {len(members)} member(s):", ""]

            for i, node in enumerate(members, 1):
                title = node.get("content_title", "Untitled")
                ntype = node.get("subtype", node.get("type", "?"))
                if isinstance(ntype, str) and ntype.startswith("custom:"):
                    ntype = ntype[7:]
                date = str(node.get("provenance_created_at", "?"))[:10]
                nid = node.get("id", "?")
                lifecycle = node.get("state_lifecycle", "")

                lifecycle_tag = (
                    f" [{lifecycle}]"
                    if lifecycle and lifecycle != "ACTIVE"
                    else ""
                )
                lines.append(
                    f"  {i}. [{ntype}] {title} ({date}, {nid}){lifecycle_tag}"
                )

                # Preview
                body = node.get("content_body", "") or ""
                summary = node.get("content_summary", "") or ""
                preview = ""
                if body.startswith("{"):
                    try:
                        parsed = json.loads(body)
                        preview = parsed.get("text", "")
                    except (json.JSONDecodeError, TypeError):
                        pass
                if not preview:
                    preview = body or summary
                if preview:
                    if len(preview) > 150:
                        preview = preview[:147] + "..."
                    lines.append(f"     {preview}")

            return "\n".join(lines)

        # ---- HEALTH ----
        elif action == "health":
            if not cluster_id:
                return "Error: cluster_id is required for health action."

            health = client.get_cluster_health(cluster_id)

            name_str = health.get("name", cluster_id)
            total = health.get("total_nodes", 0)
            active = health.get("active_nodes", 0)
            weak = health.get("weak_nodes", 0)
            dormant = health.get("dormant_nodes", 0)
            ratio = health.get("health_ratio", 0)
            pct = int(ratio * 100)

            return (
                f"{name_str} ({cluster_id}) — Health: {pct}%\n"
                f"  Total: {total} | Active: {active} | "
                f"Weak: {weak} | Dormant: {dormant}"
            )

        else:
            return (
                f"Error: unknown action '{action}'. "
                "Use list, create, update, delete, add, remove, members, or health."
            )

    except Exception as e:
        logger.error("manage_clusters failed: %s", e)
        return f"Error: {e}"


def register(registry):
    """Register manage_clusters tool."""
    from .registry import Tool

    registry.register(Tool(
        name=TOOL_DEF["name"],
        description=TOOL_DEF["description"],
        parameters=TOOL_DEF["parameters"],
        handler=handle_manage_clusters,
    ))

File: nous-server/server/src/routes/clusters.ts
import { Hono } from 'hono';
import { getDb } from '../db.js';
import { clusterId, now } from '../utils.js';
import { applyDecay, type NodeRow } from '../core-bridge.js';

const clusters = new Hono();

/**
 * Parse auto_subtypes from JSON string to array.
 * SQLite stores TEXT; we return a proper array to clients.
 */
function parseAutoSubtypes(row: any): any {
  if (row && row.auto_subtypes && typeof row.auto_subtypes === 'string') {
    try {
      row.auto_subtypes = JSON.parse(row.auto_subtypes);
    } catch {
      // Malformed JSON — leave as string
    }
  }
  return row;
}

// ---- CREATE ----
clusters.post('/clusters', async (c) => {
  const body = await c.req.json();
  const { name, description, icon, pinned, auto_subtypes } = body;

  if (!name || typeof name !== 'string') {
    return c.json({ error: 'name is required' }, 400);
  }

  const id = clusterId();
  const ts = now();
  const autoSubtypesJson = Array.isArray(auto_subtypes)
    ? JSON.stringify(auto_subtypes)
    : null;

  const db = getDb();
  await db.execute({
    sql: `INSERT INTO clusters (id, name, description, icon, pinned, source, auto_subtypes, created_at)
          VALUES (?, ?, ?, ?, ?, 'user_created', ?, ?)`,
    args: [
      id, name, description ?? null, icon ?? null,
      pinned ? 1 : 0, autoSubtypesJson, ts,
    ],
  });

  const result = await db.execute({
    sql: 'SELECT * FROM clusters WHERE id = ?',
    args: [id],
  });
  return c.json(parseAutoSubtypes({ ...result.rows[0] }), 201);
});

// ---- LIST (with node counts) ----
clusters.get('/clusters', async (c) => {
  const db = getDb();
  const result = await db.execute({
    sql: `SELECT c.*, COUNT(cm.node_id) as node_count
          FROM clusters c
          LEFT JOIN cluster_memberships cm ON cm.cluster_id = c.id
          GROUP BY c.id
          ORDER BY c.pinned DESC, c.created_at DESC`,
    args: [],
  });

  const data = result.rows.map((row) => parseAutoSubtypes({ ...row }));
  return c.json({ data, count: data.length });
});

// ---- GET BY ID ----
clusters.get('/clusters/:id', async (c) => {
  const id = c.req.param('id');
  const db = getDb();
  const result = await db.execute({
    sql: `SELECT c.*, COUNT(cm.node_id) as node_count
          FROM clusters c
          LEFT JOIN cluster_memberships cm ON cm.cluster_id = c.id
          WHERE c.id = ?
          GROUP BY c.id`,
    args: [id],
  });

  if (result.rows.length === 0) return c.json({ error: 'not found' }, 404);
  return c.json(parseAutoSubtypes({ ...result.rows[0] }));
});

// ---- UPDATE (PATCH) ----
clusters.patch('/clusters/:id', async (c) => {
  const id = c.req.param('id');
  const body = await c.req.json();
  const db = getDb();

  const existing = await db.execute({
    sql: 'SELECT * FROM clusters WHERE id = ?',
    args: [id],
  });
  if (existing.rows.length === 0) return c.json({ error: 'not found' }, 404);

  const sets: string[] = [];
  const args: (string | number | null)[] = [];

  // Simple fields
  for (const field of ['name', 'description', 'icon'] as const) {
    if (field in body) {
      sets.push(`${field} = ?`);
      args.push(body[field] ?? null);
    }
  }

  // Boolean field — stored as INTEGER
  if ('pinned' in body) {
    sets.push('pinned = ?');
    args.push(body.pinned ? 1 : 0);
  }

  // JSON field — serialize array to string
  if ('auto_subtypes' in body) {
    sets.push('auto_subtypes = ?');
    args.push(
      Array.isArray(body.auto_subtypes)
        ? JSON.stringify(body.auto_subtypes)
        : null,
    );
  }

  if (sets.length === 0) {
    return c.json({ error: 'no valid fields to update' }, 400);
  }

  args.push(id);
  await db.execute({
    sql: `UPDATE clusters SET ${sets.join(', ')} WHERE id = ?`,
    args,
  });

  const result = await db.execute({
    sql: 'SELECT * FROM clusters WHERE id = ?',
    args: [id],
  });
  return c.json(parseAutoSubtypes({ ...result.rows[0] }));
});

// ---- DELETE (manual cascade — remove memberships first) ----
clusters.delete('/clusters/:id', async (c) => {
  const id = c.req.param('id');
  const db = getDb();
  await db.execute({
    sql: 'DELETE FROM cluster_memberships WHERE cluster_id = ?',
    args: [id],
  });
  await db.execute({
    sql: 'DELETE FROM clusters WHERE id = ?',
    args: [id],
  });
  return c.json({ ok: true });
});

// ---- ADD MEMBER(S) ----
clusters.post('/clusters/:id/members', async (c) => {
  const id = c.req.param('id');
  const body = await c.req.json();
  const { node_id, node_ids } = body;

  const db = getDb();

  // Verify cluster exists
  const cluster = await db.execute({
    sql: 'SELECT id FROM clusters WHERE id = ?',
    args: [id],
  });
  if (cluster.rows.length === 0) {
    return c.json({ error: 'cluster not found' }, 404);
  }

  // Accept single node_id or array node_ids
  const ids: string[] = node_ids ?? (node_id ? [node_id] : []);
  if (ids.length === 0) {
    return c.json({ error: 'node_id or node_ids is required' }, 400);
  }

  const ts = now();
  let added = 0;

  for (const nid of ids) {
    try {
      await db.execute({
        sql: `INSERT OR IGNORE INTO cluster_memberships
              (cluster_id, node_id, weight, pinned, updated_at)
              VALUES (?, ?, 1.0, 0, ?)`,
        args: [id, nid, ts],
      });
      added++;
    } catch {
      // Node doesn't exist or other constraint — skip silently
    }
  }

  return c.json({ ok: true, added, cluster_id: id });
});

// ---- REMOVE MEMBER ----
clusters.delete('/clusters/:id/members/:node_id', async (c) => {
  const id = c.req.param('id');
  const nodeId = c.req.param('node_id');
  const db = getDb();
  await db.execute({
    sql: 'DELETE FROM cluster_memberships WHERE cluster_id = ? AND node_id = ?',
    args: [id, nodeId],
  });
  return c.json({ ok: true });
});

// ---- LIST MEMBERS ----
clusters.get('/clusters/:id/members', async (c) => {
  const id = c.req.param('id');
  const limit = Math.min(Number(c.req.query('limit') ?? 50), 200);
  const db = getDb();

  const result = await db.execute({
    sql: `SELECT n.*,
            cm.weight as membership_weight,
            cm.updated_at as membership_updated_at
          FROM cluster_memberships cm
          JOIN nodes n ON n.id = cm.node_id
          WHERE cm.cluster_id = ?
          ORDER BY cm.updated_at DESC
          LIMIT ?`,
    args: [id, limit],
  });

  // Apply FSRS decay for display (same pattern as GET /nodes list)
  const data = result.rows.map((row) =>
    applyDecay(row as unknown as NodeRow),
  );
  return c.json({ data, count: data.length });
});

// ---- HEALTH ----
clusters.get('/clusters/:id/health', async (c) => {
  const id = c.req.param('id');
  const db = getDb();

  // Verify cluster exists
  const cluster = await db.execute({
    sql: 'SELECT id, name FROM clusters WHERE id = ?',
    args: [id],
  });
  if (cluster.rows.length === 0) {
    return c.json({ error: 'not found' }, 404);
  }

  // Count nodes per lifecycle state
  const stats = await db.execute({
    sql: `SELECT
            COUNT(*) as total_nodes,
            SUM(CASE WHEN n.state_lifecycle = 'ACTIVE' THEN 1 ELSE 0 END) as active_nodes,
            SUM(CASE WHEN n.state_lifecycle = 'WEAK' THEN 1 ELSE 0 END) as weak_nodes,
            SUM(CASE WHEN n.state_lifecycle = 'DORMANT' THEN 1 ELSE 0 END) as dormant_nodes
          FROM cluster_memberships cm
          JOIN nodes n ON n.id = cm.node_id
          WHERE cm.cluster_id = ?`,
    args: [id],
  });

  const row = stats.rows[0] as any;
  const total = Number(row.total_nodes ?? 0);
  const active = Number(row.active_nodes ?? 0);

  return c.json({
    cluster_id: id,
    name: (cluster.rows[0] as any).name,
    total_nodes: total,
    active_nodes: active,
    weak_nodes: Number(row.weak_nodes ?? 0),
    dormant_nodes: Number(row.dormant_nodes ?? 0),
    health_ratio: total > 0 ? Math.round((active / total) * 10000) / 10000 : 0,
  });
});

export default clusters;
